### PR TITLE
Exceptions in the proxy_protocol.cc file have been replaced 

### PR DESCRIPTION
**Pull request by mrSeoD**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message: In order to replace throwing exceptions in the proxy_protocol.cc file, three different ways were implemented: return enum values for methods that were originally assigned to be return "void" (onReadWorker(), parseV1Header(), parseV2Header()), return false and print out the log statements for the methods that were originally assigned to return booleans (parseExtensions(), readProxyHeader()), and return nullopt for a method (lenV2Address()) that was originally designed to return a size_t values. onRead() and readProxyHeader() methods call these methods, so they are designed to break out of their own methods as soon as an error occurs in the methods that they called. 

Additional Description: This pull request is regarding to the issue #11044

Risk Level: Low-Medium

Testing: As a beginner-contributor, I apologize for not being able to test this code. I am not sure how to test it. Any guidance/recommendation on testing these changes would be amazing for me to make sure the code works well. @asraa has been working with me for awhile and we thought PRing the code would be faster and more efficient at this time to fix the problems. 

Docs Changes: N/A
Release Notes: N/A
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by asraa** [comment]
Thanks! Will take a look. In the meantime can you please take a look at these docs to fix DCO (https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#fixing-dco) and look over the README to fix formatting issues? https://github.com/envoyproxy/envoy/blob/master/support/README.md

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc [hunk 1]
```diff
@@ -284,48 +298,63 @@ bool Filter::readProxyHeader(os_fd_t fd) {
...
         if (read_result.rc_ != exp) {
-          throw EnvoyException("failed to read proxy protocol (remote closed)");
+            ENVOY_LOG(debug, "failed to read proxy protocol (remote closed)");
+            return false;
         }
         buf_off_ += read_result.rc_;
         nread -= read_result.rc_;
       }
-      ssize_t addr_len = lenV2Address(buf_);
+      ssize_t addr_len = lenV2Address(buf_).value_or(NO_VALUE_ASSIGNED);
```
Comment (asraa) [inline source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:322]: Why not test against `has_value()`? That way you can remove `NO_VALUE_ASSIGNED`

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.h [hunk 2]
```diff
@@ -43,6 +43,17 @@ using ConfigSharedPtr = std::shared_ptr<Config>;
 
 enum ProxyProtocolVersion { Unknown = -1, InProgress = -2, V1 = 1, V2 = 2 };
 
+enum class ProxyState{
+    //failed to read proxy protocol
+    PrtclFail,
+
+    //Unsupported command or address family or transport
```
Comment (asraa) [inline source/extensions/filters/listener/proxy_protocol/proxy_protocol.h:50]: Can you document what the usage of this is? it looks like it is only used in one method, and the try/catch that's replaced only tests for `PrtclFail`.

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.h [hunk 3]
```diff
@@ -43,6 +43,17 @@ using ConfigSharedPtr = std::shared_ptr<Config>;
 
 enum ProxyProtocolVersion { Unknown = -1, InProgress = -2, V1 = 1, V2 = 2 };
 
+enum class ProxyState{
+    //failed to read proxy protocol
+    PrtclFail,
```
Comment (asraa) [inline source/extensions/filters/listener/proxy_protocol/proxy_protocol.h:48]: Can you give this a name without abbreviations for clarity? `ProtocolFail`? `Error`?

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.h [hunk 4]
```diff
@@ -43,6 +43,17 @@ using ConfigSharedPtr = std::shared_ptr<Config>;
...
+
+    //Unsupported command or address family or transport
+    UnsupportedCd,
+
+    //the method's job is over
```
Comment (asraa) [inline source/extensions/filters/listener/proxy_protocol/proxy_protocol.h:53]: Please fix format of comments, add a space after '/' and a period please. I think fix format script will do this.

Comment (mrSeoDavid) [reply source/extensions/filters/listener/proxy_protocol/proxy_protocol.h:53]: I ran the lines in the link you've provided; however, they did not work; so I've manually fixed them.

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.h (hunk 3 above)
Comment (mrSeoDavid) [reply source/extensions/filters/listener/proxy_protocol/proxy_protocol.h:48]: I've changed them to "ProtocolFail" in both .h and .cc files

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc (hunk 1 above)
Comment (mrSeoDavid) [reply source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:322]: Perfect!

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.h (hunk 2 above)
Comment (mrSeoDavid) [reply source/extensions/filters/listener/proxy_protocol/proxy_protocol.h:50]: I believe ProtocolFail (changed it from PrtclFail following your next comment) enum value is used in multiple methods like onRead(), onReadWorker(), parseV1Header, and so on. OR, are you referring to something else? 

The try/catch that's replaced only tests for ProtocolFail as it was supposed to throw an exception if there was this error only. Is there any change that you would like to recommend?

**Comment by mrSeoDavid** [comment]
@asraa Thank you so much for your help! Really appreciate it. 

For the DCO error, I ran this: 

git rebase -i HEAD^^
(interactive squash + DCO append)
git push origin -f

However, it seems like the DCO error is not going away.. 
____________________________________________________________________________________
Also, for the new enum values that I've created like ProtocolFail, I've realized that there was an enum class already called "enum ProxyProtocolVersion." Should I go ahead and just put these newly-created enum values by me into this class and delete the "enum class ProxyState"?
____________________________________________________________________________________
Lastly, I tried pushing new changes to the master branch by:

git checkout master
git push origin

However, it seems like there were no updates.. Any help would be amazing

Thank you so much once again!

**Comment by mrSeoDavid** [comment]
@asraa I was able to make fix the DCO problem and was able to push new changes on to this master branch!

Please disregard my first and third question in the above comment, but only consider the second question. If there is anything that I need to fix, please let me know! Thanks a lot.

**Comment by mrSeoDavid** [comment]
@asraa Also a quick general question, when I committed a new change and pushed  8aef413, the changes that were made successfully got uploaded; however, the codes that I've fixed on my local computer automatically changed back to the previous code. Do you know why this is happening and how to fix this? Thanks a lot again.

**Comment by mrSeoDavid** [comment]
@asraa I detected some problems and had to fix it again and commit&push the codes again. Now the codes DO NOT go back to the previous codes; however, now the DCO problem is not being solved. If I run: 

"git rebase -i HEAD~2" OR "git rebase -i HEAD^^"  BOTH will give this message: 
noop
 
# Rebase b178f6ed6..b178f6ed6 onto b178f6ed6 (1 command)
#
# Commands:
# p, pick <commit> = use commit
# r, reword <commit> = use commit, but edit the commit message
# e, edit <commit> = use commit, but stop for amending
# s, squash <commit> = use commit, but meld into previous commit
# f, fixup <commit> = like "squash", but discard this commit's log message
# x, exec <command> = run command (the rest of the line) using shell
# b, break = stop here (continue rebase later with 'git rebase --continue')
# d, drop <commit> = remove commit
# l, label <label> = label current HEAD with a name
# t, reset <label> = reset HEAD to a label
# m, merge [-C <commit> | -c <commit>] <label> [# <oneline>]
# .       create a merge commit using the original merge commit's
# .       message (or the oneline, if no original merge commit was
# .       specified). Use -c <commit> to reword the commit message.
#
# These lines can be re-ordered; they are executed from top to bottom.
#
</seoda/envoy/.git/rebase-merge/git-rebase-todo [unix] (16:55 12/05/2020)1,1 Top


and I would not be able to take any further action.

**Comment by mrSeoDavid** [comment]
@asraa I've figured out the DCO problem again by committing and pushing one last time. Finally, phew. 👍 

Please check fda621c to check the changes that were made following your comments as cf2a76a was just changes only for fixing DCO problem. Really appreciate your help.

File: ci/repokitteh/modules/ownerscheck.star [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+  _reconcile(config, specs)
+
+
+handlers.pull_request(func=_pr)
+handlers.pull_request_review(func=_pr_review)
+
+handlers.command(name='checkowners', func=_reconcile)
+handlers.command(name='checkowners!', func=_force_reconcile_and_comment)
+handlers.command(name='lgtm', func=_lgtm_by_comment)
```
Comment (asraa) [inline ci/repokitteh/modules/ownerscheck.star:224]: Why is this file included? Maybe merge master?

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc [hunk 6]
```diff
@@ -42,23 +42,23 @@ Network::FilterStatus Filter::onAccept(Network::ListenerFilterCallbacks& cb) {
...
 void Filter::onRead() {
-  try {
-    onReadWorker();
-  } catch (const EnvoyException& ee) {
-    config_->stats_.downstream_cx_proxy_proto_error_.inc();
-    cb_->continueFilterChain(false);
-  }
+    ProxyState status = onReadWorker();
+
+    if(state == ProxyState::ProtocolFail){
```
Comment (asraa) [inline source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:47]: stats -> status?

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc [hunk 7]
```diff
@@ -181,14 +184,15 @@ void Filter::parseV2Header(char* buf) {
...
       }
     }
   }
-  throw EnvoyException("Unsupported command or address family or transport");
+  ENVOY_LOG(debug, "Unsupported command or address family or transport");
+  return ProxyState::UnsupportedCd;
 }
 
-void Filter::parseV1Header(char* buf, size_t len) {
+ProxyState Filter::parseV1Header(char* buf, size_t len) {
```
Comment (asraa) [inline source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:195]: The type signature for this in the header must be changed.

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc [hunk 8]
```diff
@@ -109,12 +111,13 @@ size_t Filter::lenV2Address(char* buf) {
...
   default:
-    throw EnvoyException("Unsupported V2 proxy protocol address family");
+      ENVOY_LOG(debug, "Unsupported V2 proxy protocol address family");
+      return std::nullopt;
   }
   return len;
 }
 
-void Filter::parseV2Header(char* buf) {
+ProxyState Filter::parseV2Header(char* buf) {
```
Comment (asraa) [inline source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:120]: Change function signature here as well, and others.

File: ci/repokitteh/modules/ownerscheck.star (hunk 5 above)
Comment (mrSeoDavid) [reply ci/repokitteh/modules/ownerscheck.star:224]: I think so.. not sure why it got added to the file, should I delete the "modules" folder?

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc (hunk 6 above)
Comment (mrSeoDavid) [reply source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:47]: I will make sure to run the test before committing!

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc (hunk 7 above)
Comment (mrSeoDavid) [reply source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:195]: I am not sure what change you are asking. Do you want the return-type of the method to be changed? Thanks!

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc (hunk 6 above)
Comment (mrSeoDavid) [reply source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:47]: I tried running the bazel tests however the git bash returns: "bash: bazel: command not found" Thanks for your help always.

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc (hunk 7 above)
Comment (asraa) [reply source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:195]: Yes, you changed it by making `void Filter::parserV1Header` to `ProxyState Filter::parseV1Header`. But you did not change that in the header file.

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc (hunk 6 above)
Comment (asraa) [reply source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:47]: You need to follow the build instructions to develop envoy https://github.com/envoyproxy/envoy/blob/master/bazel/README.md

If you don't want to, you can always use docker. https://github.com/envoyproxy/envoy/tree/master/ci#building-and-running-tests-as-a-developer

File: ci/repokitteh/modules/ownerscheck.star (hunk 5 above)
Comment (asraa) [reply ci/repokitteh/modules/ownerscheck.star:224]: merge master would fix it. the file exists upstream, maybe you rebased?

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc (hunk 7 above)
Comment (mrSeoDavid) [reply source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:195]: Got it! fixed everything on the .h file!

File: ci/repokitteh/modules/ownerscheck.star (hunk 5 above)
Comment (mrSeoDavid) [reply ci/repokitteh/modules/ownerscheck.star:224]: yea I did rebase.. haha thanks!

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc (hunk 6 above)
Comment (mrSeoDavid) [reply source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:47]: Thank you so much for your help!! @asraa  You are helping me learn the open source contribution so much!

**Comment by mrSeoDavid** [comment]
@asraa I fixed all the header return types. I tried to download the bazel and run for like 30 mins but it's not working.. 
Also, what does it mean by "save this binary in a directory on the PATH as bazel.exe."?
After doing this, do I only need to run the "bazel test test/extensions/filters/listener/proxy_protocol/..."?

I really appreciate your help.

**Comment by asraa** [comment]
> After doing this, do I only need to run the "bazel test test/extensions/filters/listener/proxy_protocol/..."?

I would start with that, but it's generally good practice to run the whole test suite (test/integration is another good place to start)

Are you using Windows?

**Comment by mrSeoDavid** [comment]
@asraa I'm sorry for a late reply, but yes I am using windows! Unfortunately, I was not able to download and implement bazel. The command "set PATH=%PATH%;%USERPROFILE%\bazel" is stating that no such job is possible. I tried different setting up method, but the commands are saying "no such job" exists. Any help would be appreciated!!

**Comment by asraa** [comment]
Are you in the Envoy slack channel? People using Windows may be able to help (I personally don't). 
I'm assuming you were following the instructions at https://github.com/envoyproxy/envoy/tree/master/bazel

You may also want to file a GitHub issue about it, unless you find a related GitHub issue about windows build that you want to reopen/comment on

**Comment by mrSeoDavid** [comment]
@asraa Thank you thank you so much for your help always! I will try to solve this issue and hopefully get back to you soon!
